File: converters/turn_mzml_sqlite.R

library(RaMS)
library(DBI)
library(RSQLite)
library(data.table)
library(xml2)

turn_mzml_sqlite <- function(ms_files, outfile, ordered=NULL){
  msdata <- grabMSdata(ms_files)
  
  simple_cols <- setdiff(names(msdata$metadata), "config_data")
  file_info <- msdata$metadata[,simple_cols,with = FALSE]
  
  ms_levels <- lapply(ms_files, grabAccessionData, "MS:1000511")
  ms_file_vec <- rep(basename(ms_files), sapply(ms_levels, nrow))
  ms_levels <- do.call(what = "rbind", ms_levels)
  colnames(ms_levels) <- c("name", "ms_level")
  ms_levels$filename <- ms_file_vec
  ms_levels$ms_level <- as.numeric(ms_levels$ms_level)
  ms_levels <- as.data.table(ms_levels)
  
  getScanNum <- function(filename){
    xml_data <- read_xml(filename)
    cont_nums <- xml_attr(xml_find_all(xml_data, "//d1:spectrum"), "id")
    scan_nums <- as.numeric(gsub(".*scan=", "", cont_nums))
    data.table(filename=basename(filename), scan_num=scan_nums)
  }
  scan_nums <- lapply(ms_files, getScanNum)
  scan_nums <- do.call(what = "rbind", scan_nums)
  scan_levels <- cbind(ms_levels, scan_nums[,filename:=NULL])
  scan_levels[,name:=NULL]
  ms1_nums <- scan_levels[ms_level==1,"scan_num"]
  tic <- cbind(ms1_nums, msdata$TIC)
  setnames(tic, c("scan_num", "rt", "TIC", "filename"))
  bpc <- cbind(ms1_nums, msdata$BPC)
  setnames(bpc, c("scan_num", "rt", "BPC", "filename"))
  scan_info <- merge(scan_levels, merge(tic, bpc), all.x = TRUE)
  
  conn <- dbConnect(RSQLite::SQLite(), outfile)
  dbWriteTable(conn, "file_info", file_info)
  dbWriteTable(conn, "scan_info", scan_info)
  dbWriteTable(conn, "MS1", msdata$MS1)
  dbWriteTable(conn, "MS2", msdata$MS2)
  
  if(!is.null(ordered)){
    if(ordered=="mz"){
      dbExecute(conn, "CREATE INDEX mz_idx ON MS1 (mz)")
      dbExecute(conn, "CREATE INDEX premz_idx ON MS2 (premz)")
    }
    if(ordered=="rt"){
      dbExecute(conn, "CREATE INDEX rt_idx ON scan_info (rt)")
      dbExecute(conn, "CREATE INDEX rt_idx ON MS1 (rt)")
      dbExecute(conn, "CREATE INDEX rt_idx ON MS2 (rt)")
    }
  }
  
  dbDisconnect(conn)
  return(outfile)
}

# Example usage:
ms_files <- list.files("demo_data", pattern = "mzML", full.names = TRUE)
turn_mzml_sqlite(ms_files, "demo_data/msdata.sqlite", ordered = "mz")

# Example chromatogram extraction
library(DBI)
pmppm <- function(mass, ppm)c(mass * (1 - ppm/1e+06), mass * (1 + ppm/1e+06))

conn <- dbConnect(RSQLite::SQLite(), "demo_data/msdata.sqlite")
chr_statement <- "SELECT * FROM MS1 WHERE mz BETWEEN ? AND ? ORDER BY filename, rt"
chrom <- dbGetQuery(conn, chr_statement, params=pmppm(118.0865, 10))
dbDisconnect(conn)
plot(chrom$rt, chrom$int, type = "l")

File: converters/turn_mzml_sqlite.py
import sqlite3
import pandas as pd
import pyteomics.mzml
import os
def pmppm(mz, ppm):
    return(mz*(1-ppm/1000000), mz*(1+ppm/1000000))

def turn_mzml_sqlite(files, outfile, ordered=None):
    conn = sqlite3.connect(outfile)
    conn.execute("DROP TABLE IF EXISTS MS1")
    conn.execute("DROP TABLE IF EXISTS MS2")
    if isinstance(files, str):
        files = [files]
    for file in files:
        open_file = pyteomics.mzml.MzML(file)
        file_info = []
        n_scans = len(open_file)
        min_rt = open_file[0]["scanList"]["scan"][0]["scan start time"]
        max_rt = open_file[n_scans-1]["scanList"]["scan"][0]["scan start time"]
        file_info.append(pd.DataFrame({'filename':[os.path.basename(file)], 'n_scans':n_scans,
                                       'rt_start':min_rt, 'rt_end':max_rt}))
        # Might need to switch to pymzml for more metadata...
    
        scan_info = []
        MS1_dfs = []
        MS2_dfs = []
        for spectrum in open_file:
            rt_val = spectrum['scanList']['scan'][0]['scan start time']
            idx = int(spectrum['id'].split("scan=")[-1].split()[0])
            scan_info.append(pd.DataFrame({'filename':[os.path.basename(file)], 'rt':rt_val,
                                           'scan_num':idx, 'min_mz':spectrum['lowest observed m/z'], 
                                           'max_mz':spectrum['highest observed m/z'],
                                           'ms_level':spectrum["ms level"],
                                           'centroided':'centroid spectrum' in demo_lib,
                                           'polarity':("negative", "positive")['positive scan' in demo_lib]}))
            
            if spectrum['ms level'] == 1:
                mz_vals = spectrum['m/z array']
                int_vals = spectrum['intensity array']
                df_scan = pd.DataFrame({'filename': os.path.basename(file), 'id': idx, 'mz': mz_vals, 
                                        'int': int_vals, 'rt': [rt_val] * len(mz_vals)})
                MS1_dfs.append(df_scan)
            if spectrum['ms level'] == 2:
                mz_vals = spectrum['m/z array']
                int_vals = spectrum['intensity array']
                premz_val = spectrum['precursorList']['precursor'][0]['isolationWindow']['isolation window target m/z']
                df_scan = pd.DataFrame({'filename': os.path.basename(file), 'id': idx, 'premz': premz_val, 
                                        'fragmz': mz_vals, 'int': int_vals, 'rt': [rt_val] * len(mz_vals)})
                MS2_dfs.append(df_scan)
        pd.concat(file_info, ignore_index=True).to_sql("file_info", conn, if_exists="append", index=False)
        pd.concat(scan_info, ignore_index=True).to_sql("scan_info", conn, if_exists="append", index=False)
        pd.concat(MS1_dfs, ignore_index=True).to_sql("MS1", conn, if_exists="append", index=False)
        pd.concat(MS2_dfs, ignore_index=True).to_sql("MS2", conn, if_exists="append", index=False)
    if ordered is not None:
        index_name = f"idx_{ordered}"
        if ordered == "rt":
            conn.execute(f"CREATE INDEX IF NOT EXISTS {index_name} ON MS1 ({ordered})")
            conn.execute(f"CREATE INDEX IF NOT EXISTS {index_name} ON MS2 ({ordered})")
        if ordered == "mz":
            conn.execute(f"CREATE INDEX IF NOT EXISTS {index_name} ON MS1 ({ordered})")
        if ordered in ["fragmz", "premz"]:
            conn.execute(f"CREATE INDEX IF NOT EXISTS {index_name} ON MS2 ({ordered})")
    conn.close()
    
    return outfile



import glob
ms_files = glob.glob("demo_data/*.mzML")
turn_mzml_sqlite(ms_files, "demo_data/msdata.sqlite")

# Example chromatogram extraction
conn = sqlite3.connect("demo_data/msdata.sqlite")
chr_statement = "SELECT * FROM MS1 WHERE mz BETWEEN ? AND ? ORDER BY filename, rt"
chrom = pd.read_sql(chr_statement, conn, params=pmppm(118.0865, 10))
conn.close()
import matplotlib.pyplot as plt
plt.plot(chrom["rt"], chrom["int"])
plt.show()
